### PR TITLE
core(full-page-screenshot): use layoutViewport width

### PR DESCRIPTION
**Pull request by connorjclark**
found while investigating https://github.com/GoogleChrome/lighthouse/issues/11118

You can test the before/after on http://localhost:10200/perf/trace-elements.html

note that the image doesn't fit the screen size–
image this means the `scrollWidth` of the html element is 488, but we expected the width to be 360 (width of moto g4). This resulted in the x-axis being mis-scaled in rendering the thumbnails.

before (note the highlight rectangle should cover the width of the page, but doesnt)

image

after

image

### CODE REVIEW

File: lighthouse-core/gather/gatherers/full-page-screenshot.js [hunk 1]
```diff
@@ -32,7 +32,12 @@ class FullPageScreenshot extends Gatherer {
     const driver = passContext.driver;
     const metrics = await driver.sendCommand('Page.getLayoutMetrics');
     const deviceScaleFactor = await driver.evaluateAsync('window.devicePixelRatio');
-    const width = Math.min(metrics.contentSize.width, maxScreenshotHeight);
+
+    // Emulate page with the same width, and height set to content height.
+    // Easiest way to use the same width is metrics.layoutViewport–there is also
+    // metrics.visualViewport which accounts for zooming.
+    // Note: If the page is zoomed, many assumptions fail.
```
Comment (paulirish) [inline lighthouse-core/gather/gatherers/full-page-screenshot.js:39]:
```suggestion
    // Width should match emulated width, without considering content overhang. 
    //   Both layoutViewport and visualViewport capture this. visualViewport accounts for page zoom/scale, which we currently don't account for (or expect). So we use layoutViewport.width
    // Note: If the page is zoomed, many assumptions fail.
    // Height should be as tall as the content. So we use contentSize.height
```

File: lighthouse-core/test/gather/gatherers/full-page-screenshot-test.js [hunk 2]
```diff
@@ -41,6 +41,7 @@ function createMockDriver({contentSize, screenSize, screenshotData}) {
       if (method === 'Page.getLayoutMetrics') {
         return {
           contentSize,
+          layoutViewport: {clientWidth: contentSize.width, clientHeight: contentSize.height},
```
Comment (paulirish) [inline lighthouse-core/test/gather/gatherers/full-page-screenshot-test.js:44]:
```suggestion
          // See comment within _takeScreenshot() implementation
          layoutViewport: {clientWidth: contentSize.width, clientHeight: contentSize.height},
```

**Review by paulirish — APPROVED** [review]

**Comment by paulirish** [comment]
you got a smoke failure. at first i thought it was a flake.. https://github.com/GoogleChrome/lighthouse/pull/11428



but its not.


under emulation this page looks like this

image

funnily the "full-size screenshot" that devtools takes looks like this:

image

as we know.. devtools still uses content size.. and you can tell by the TINY lil text `<p>` that this screenshot is WAY too wide, given the original emulation.

so..

the smoke failure is actually indiciative that the 5000px attempt to try to get a TOO HUGE data uri does't work because we're taking a much smaller screenshot:

image


1440px instead of 5000px.

and thus the datauri comes in under the threshold.

yay.

**Comment by connorjclark** [comment]
It's passing locally for me. you?

**Comment by paulirish** [comment]
> It's passing locally for me. you?

`yarn smoke screenshot` **fails** locally for me on this branch. 

(though since screenshot size is dependent on DPR, i wouldn't be surprised if i locally had diff results than the CI bot)

> 
> > i still think my changes in #11428 are good.. wanna pull them in here?
> 
> your changes are just deleting the test?

1. it adds an assertion in the already-existing too-large unit test. the test already asserts a null return value (which is colocated with populating a warning). The newly added assertion just doublechecks the warning is there.
2. it deletes the smoketest. the smoketest only checks that the artifact is null and runwarning is issued. nothing fancy going on here, and this behavior is already validated by the unit test. i don't see a reason to keep an entire smoketest to assert that a `if (num >= maxnum)` conditional works when that's already checked in the unit test. but maybe i'm missing something.

**Comment by patrickhulce** [comment]
> i don't see a reason to keep an entire smoketest to assert that a if (num >= maxnum) conditional works when that's already checked in the unit test. but maybe i'm missing something.

While the specific too big error case isn't necessarily super critical. The general flow of handling too large screenshots in a non-super-fatal way is very compelling to me. Like with most gatherers, the chance for failure is unlikely to be in our specific internal ifs and far more likely in a misinterpretation or expectation of how the integration with the protocol actually works. And the failure mode for pages that are too tall is both significant and very common. Any tall page on desktop easily sails passed 2MB. If we convert to a test on the handling a tall page by cutting down to <5000px tall instead of the error that seems fine to me but *something* in smokes to cover this too big handling case seems important.

**Comment by paulirish** [comment]
> And the failure mode for pages that are too tall is both significant and very common

there's two situations here and i'm not sure which you're talking about.. perhaps both?

* page too tall: chrome can't take a fullpage screenshot that's taller than 16384px (actually 16383). it'll get one that's that tall, but it wont include any content past that line.
* image too big: apparently a max size for data uris of 2MB. 


> If we convert to a test on the handling a tall page by cutting down to <5000px tall instead of the error to reduce flakes that seems fine to me but _something_ in smokes to cover this too big handling case seems important.

are you saying our test would assert that we reattempted at 5000px? and it worked or didnt or both?

-------------

On the data uri max size.. I see it in our code, the chromium bug [comments](https://bugs.chromium.org/p/chromium/issues/detail?id=69227), and [the chromium source](https://source.chromium.org/chromium/chromium/src/+/master:url/url_constants.cc;l=36;drc=55caf9065ea787560078379593eca648d123834e).

However... I'm not sure if this limit is a problem for us.  

I logged out the size of the data uri that devtools is creating when it makes a fullsize screenshot of [our test page](http://localhost:10200/screenshot.html?width=5000px&height=5000px)..

image

and it hits 10MB and saves just fine.

perhaps that 2MB is for top-level URLs being navigated to?

**Comment by patrickhulce** [comment]
> page too tall: chrome can't take a fullpage screenshot that's taller than 16384px (actually 16383). it'll get one that's that tall, but it wont include any content past that line.

> are you saying our test would assert that we reattempted at 5000px? and it worked or didnt or both?

I would give slight preference to a test that it worked after being reattempted, then the version we currently have, then one higher than the chrome limit but a high priority that at least one of these exists that it doesn't fail to handle the tall case entirely. (with the below, the first two might be moot)

> and it hits 10MB and saves just fine.

My understanding was that the concern isn't creating and saving, it's rendering the screenshot using that data URL. That being said, I did just open up a page with a 2.5MB data URL image and it worked (the only effect seems to be that Chrome disables the menu option for opening the image in a new tab) sooooooooo maybe neither of those first two cases are necessary at all? I guess we never double checked those constraints were necessary after the PR was revived from the dead, sorry @connorjclark 🤦

**Comment by connorjclark** [comment]
So just delete the too-big smoke test?

**Pull request merged**